### DISCUSSION TITLE
On Rover Languages

### DISCUSSION THREAD

**Discussion by rhettg** [rhettg/stubborn#3] [Ideas]
I've been thinking lately about the Command Language I use to interact with Stubborn. Specifically the commands an operator uses to specify the movement and operations of the rover.

Importantly, this is distinct from:

* The ground station UI, which is an interactive console.
* Implementation language of software running on either Rover or Ground Station.

In the current implementation, the command language for the ground station and for the rover *seem* the same. But only a subset of commands typed into the console are translated into commands for the rover. I think it should be expected that these will further diverge over time as more advanced forms of the ground station are developed.

### Current Implementation

Currently, the rover understands 3 commands as a core part of the CiC Platform:

https://github.com/rhettg/stubborn/blob/bd625b1079c95ba16ac1b010cd41ef3915e262f7/libraries/CiC/src/ci.h#L6-L8

The ground station converts a string based representation into these commands as follows:

| Command | CI           | Description                        |
|---------|--------------|------------------------------------|
| NOOP    | CI_CMD_NOOP  | No Operation. Essentially a "ping" |
| CLEAR   | CI_CMD_CLEAR | Clear any open error codes         |
| BOOM    | CI_CMD_BOOM  | Create synthetic error             |

Stubborn then defines a set of "mission specific" commands outside of the flight system. These are more specific to the abilities of the rover, though are also probably common to many use cases.

https://github.com/rhettg/stubborn/blob/bd625b1079c95ba16ac1b010cd41ef3915e262f7/libraries/Stubborn/src/stubborn.h#L7-L13

| Command | CI              | Description             |
|---------|-----------------|-------------------------|
| STOP    | CI_CMD_EXT_STOP | Stop motors             |
| FWD     | CI_CMD_EXT_FWD  | Forward, minimal speed. |
| FFWD    | CI_CMD_EXT_FFWD | Forword, max speed      |
| BCK     | CI_CMD_EXT_BCK  | Backward                |
| RT      | CI_CMD_EXT_RT   | Right Turn              |
| LT      | CI_CMD_EXT_LT   | Left Turn               |
| SET     | CI_CMD_EXT_SET  | Set tbl value.          |

Commands in general have support for an addition 32 bits of arguments. Typically they use 8 bits. 

The arguments in the form:

    -> FWD 10

Indicate the duration the command should run for. The number is multiplied by 100ms to build the full duration. So in the above example, both motors will be engaged in the forward direction for 1 second.

### Previously

In an earlier implementation, the commands were the same, but rather than specify a duration the arguments were used to control velocity.

   ->  FWD 255

This indicates moving forwards at full speed (255 being the max 8 bit value that represents throttle control).

The downsides for this scheme were the necessarily interactive needs of controlling the rover: It starts moving, and the operator needs to stop it before it runs off a cliff. This was surprisingly difficult in practice.

### Future

What I'm finding with the current implementation is that it still feels a little *too* interactive and slow. The operator types a command that moves the rover for a short period of time. Then they type the next one. 

This scheme does not scale for a hypothetical world where latency mattered and/or additional automation was desired. What could be useful is to upload a "program" of multiple instructions to accomplish some mission. A rover virtual machine.

```
FWD 
RT
FWD
RT
FWD
RT
PHOTO
LT 180
FWD
LT
FWD
LT
```

But how should control flow or even arguments work in this world? 

In the above example, the commands clearly have a semantic meaning where `FWD` indicates a velocity as well as a duration.   But if these friendly defaults are not acceptable?

```
SET SPEED 255
FWD
```

Global variables/settings are one potentially solution.

```
FWD
WAIT 10
STOP
```

In this example, sequential commands are executed immediately. A command `WAIT` allows the program to pause and continue the action. This has a certain elegance and flexibility to it. 

Going even lower level could be a direction to explore:

```
MOTOR A 255
MOTOR B 255
WAIT 10
MOTOR A 0
MOTOR B 0
```

A bit tedious, but if this is a "virtual machine", perhaps higher-level tooling "compiles" into this language.

One feature the above are all missing is dealing with feedback. If a robot is supposed to respond to sensors and make decisions about the world around it, how can this be implemented in a virtual machine like setup that can be re-programmed on the fly.

In Stubborn's case, the Impact Sensor provides a good use case. How do I control the behavior of the rover when the sensor is triggered?

### Next Steps

Lots of open questions. But I know I want to be able to "upload" a sequence of instructions to the Rover, so I can at least start there.

I could use some help with:

* Is there helpful prior art? Some of the code looks a lot like Logo. Stubborn even has a passing resemblance to the programable turtle from the 70s
<img width="398" alt="image" src="https://user-images.githubusercontent.com/50074/122861564-7acbac00-d2d4-11eb-930b-9d2a08dac204.png">
 
* What are other necessary features such a language needs?
* Has anyone ever seen any resources describing the details of "programming" Curiosity or Perseverance?
* Are there obvious solutions for more powerful platforms that I'm ignoring simply because this is a computationally constrained environment?

**Comment by ic** [comment]
Great discussion!

I have considered (and dropped due to time constraints) programming a simple joystick driver to emit the commands. This could be done on the current API and be much fun... After all the actual reasons the language feels tedious is not really the language itself, but typing the commands. Keyboard and joysticks are basically close interfaces, for specialized uses!

> One feature the above are all missing is dealing with feedback. If a robot is supposed to respond to sensors and make decisions about the world around it, how can this be implemented in a virtual machine like setup that can be re-programmed on the fly.
> 
> In Stubborn's case, the Impact Sensor provides a good use case. How do I control the behavior of the rover when the sensor is triggered?

Typical case for a [subsumption architecture](https://en.wikipedia.org/wiki/Subsumption_architecture). The basic idea is to prioritize events and commands, in a single event loop. One way to do it is to have a prioritized queue of events to be executed, where events are either from sensors or from commands (impact sensor signal and any command signal on equal footage, but different priorities). The queue is processed in FIFO except when a higher priority item comes in, which gets executed first. In SA, there are at least 2 priorities (layers in the former architecture), where top priority is given to the "lowest level" (critical to survival) to "higher level" (typically planning, much slower).

Another way, more modern perhaps is to see the program as implementing [system 1 and 2](https://www.amazon.com/Thinking-Fast-Slow-Daniel-Kahneman/dp/0374533555). But an implementation could follow the above approach...

**Reply by rhettg** [reply]
You've reminded me that my first iteration of playing with motors and an arduino had a joystick. It was satisfying, though it starts to bring up questions of what's the difference between a Rover and an RC car.

system 1 and 2 is an interesting way to model this idea. The code running at the lowest level of the rover could be thought of as the base capabilities that operate *fast*, with the "programmable" dynamic part *slow* and methodical. 

The Impact sensor, again as my only example, does happen to make a good practical example of this. The system 1 detects the value change on an analog pin and interprets it as "an impact occurred". Now the question is, does system 1 decide to do something with that information or is it up to system 2. Should stopping when the rover impacts an obstruction be automatic or part of the mission plan?

I've never seen "events" modeled in a BASIC, but I'm wondering if a model like this would be useful:

```
10: FWD
20: ON TIME 5000s 30
21: ON IMPACT 30
30: STOP
```

**Reply by tilnow** [reply]
you say `higher-level tooling "compiles"`
i would separate the two issues:
1. what can stubborn understand
2. how does rhett give instructions
i am for now more interested in item 2. logo requires the user to build the abstraction model. i am wondering what abstraction model would be useful for a range of rovers. i suspect that has to do with the essence of roving and with the actual task we want rover to be used for.
my system is "build 2-3 and look for commonalities and differences"

**Reply by ic** [reply]
> my first iteration of playing with motors and an arduino had a joystick. It was satisfying, though it starts to bring up questions of what's the difference between a Rover and an RC car.

Autonomous mobility and remote control seem powerful used together, especially if we think about reach problem, like in Space or deep sea, or even near the heart of a nuclear plant. Or even airplanes. They can land automatically for ages, as well as fly on cruise control. The pilot is in control (not remote, but can be, e.g. on drones) for some maneuvers only (I heard a decade ago taking off was manual, while landing was automated).

The difference is important to me, so as to understand how to best leverage the pair human-machine.

> Now the question is, does system 1 decide to do something with that information or is it up to system 2. Should stopping when the rover impacts an obstruction be automatic or part of the mission plan?

In SA and related, system 1 is basically the most basic and the highest priority. It means that when an input triggers system 1, the output is produced immediately (fast), and system 2 has nothing to say (no time to, anyway).

An example is, say, a cricket. Its system 2 may conduct her to some food. On the way, a suspicious sudden wind gets registered by her leg movement detector hairs, triggering an "automatic" system 1 reaction: Jump far. The evolutionary grounding is understood this is for survival. So system 2 is completely overridden, not even consulted, although system 2 can still be informed of what happened, to build strategic future plans.

So the impact sensor fits pretty well the system 1 bill on survival for the rover, and it should be a fail-safe, overriding anything that could interfere with critical survival routine.

> I've never seen "events" modeled in a BASIC, but I'm wondering if a model like this would be useful

BASIC has no notion of event (well, at least the versions I was using in the 80s and early 90s). However there may be access to some interrupts (I have never considered that back then!)

```
20: ON TIME 5000s 30
21: ON IMPACT 30
```

If `IMPACT` can be linked to an interrupt, and `TIME` to another with "lower priority", it looks like this can map to systems 1/2 kind of mechanism.

Do you have the possibility to customize your programming language? That's super exciting for exploration if you do!

**Reply by tilnow** [reply]
which raises the issue: how many levels of priority should there be? 2 makes sense. a large number does not. but perhaps 3:
a. normal
b. urgent takeover
c. urgent limited takeover - checks a bit before it leaps and consumes fewer resources.
